x509-cert: provide parsing profiles

baloo opened the pull request:
This allow the user to relax checks when parsing certificate and cover for non rfc5280 compliant x509 certificates.

Fixes #978
Fixes #984

cc @lumag

tarcieri commented on x509-cert/src/certificate.rs:
```diff
@@ -3,14 +3,43 @@
 #[cfg(feature = "pem")]
 use der::pem::PemLabel;
 
+pub trait ParsingProfile: PartialEq + Debug + Eq + Clone {
```
I assume `ParsingProfile` is to disambiguate from other potential usages of the word "profile", but it seems a little unwieldy, especially when all of the generic parameters are named `PP`.

baloo replied:
Yeah, mostly because I was using `Profile` in the builder PR actually :D
I'll rename that.

baloo commented on x509-cert/tests/certificate.rs:
```diff
@@ -116,7 +117,7 @@ fn reencode_cert() {
     let defer_cert = DeferDecodeCertificate::from_der(der_encoded_cert).unwrap();
 
-    let parsed_tbs = TbsCertificate::from_der(defer_cert.tbs_certificate).unwrap();
+    let parsed_tbs = TbsCertificate::<Rfc5280>::from_der(defer_cert.tbs_certificate).unwrap();
```
I don't get why I need to specify the profile here. I feel like I'm missing something.

tarcieri replied:
Default generic parameters aren't used for method invocations (or any expression). Inference is instead.

[`std` works around this by hardcoding the default parameter in certain contexts](https://doc.rust-lang.org/std/collections/struct.HashMap.html#impl-HashMap%3CK%2C%20V%2C%20RandomState%3E), although that works better for inherent methods than it does traits.

An alternative to a default generic parameter which might provide better ergonomics is to use type aliases instead.

baloo replied:
Yeah type aliases would work.

baloo commented on x509-cert/src/certificate.rs:
```diff
@@ -73,7 +108,7 @@ impl Default for Version {
 #[derive(Clone, Debug, Eq, PartialEq, Sequence, ValueOrd)]
 #[allow(missing_docs)]
-pub struct TbsCertificate {
+pub struct TbsCertificate<PP: ParsingProfile = Rfc5280> {
```
I think I like the profile to be attached to all the types better.

tarcieri replied:
There are definitely tradeoffs either way.

The main thing that bothers me about it is applying changes to the profile to new fields means you have to add a generic parameter to the type for that field, where having the logic defined on the `Profile` means it can be easily extended and modified without breaking changes.

That said, splitting it up does mean the logic is nicely factored on a type-by-type basis, instead of the profile having all of the responsibility.

tarcieri commented:
@baloo this needs a rebase

tarcieri commented on der/derive/src/attributes.rs:
```diff
@@ -79,6 +79,9 @@ pub(crate) struct FieldAttrs {
     /// Is the inner type constructed?
     pub constructed: bool,
+
+    /// Is the type skipped?
+    pub skipped: Option<Path>,
```
if this is all to handle `PhantomData`, what if instead we just impl'd the relevant `der` traits on `PhatomData` as no-ops?

baloo replied:
Yes.
I guess that could work.

baloo replied:
Except I can't implement `der::Tagged` ?

baloo replied:
I mean, this has to be handled in the `Sequence` implementation body, and thus has to be handled in the derive. I am confusing myself here?

baloo commented:
I guess I will need to split the `der/derive` changes in another PR, get those released and need a PR to fix both fuzz and the minimal-version checks

tarcieri commented:
Let me take a stab at `PhatomData` handling with derive...

baloo commented:
this is rebased on https://github.com/RustCrypto/formats/pull/1012

baloo commented:
it passes locally, I assumed the failing jobs were because we need a new patch release of der and bump the minimal-versions. (This is at least the case for minimal-versions and fuzz, I assumed the other two were cancelled because the two previous tests failed)

tarcieri commented:
This one is a straight up test failure: https://github.com/RustCrypto/formats/actions/runs/4736068365/jobs/8427760079?pr=987

baloo commented:
yes, I deleted my previous comment and trying to figure out what is happening right now.

baloo commented:
Because I dropped the handling for generics in the derive at the same time as I dropped the handling for the skipped fields.
I still need the `der_derive` to handle type generics.

tarcieri commented:
@baloo aah yeah, to properly handle generics it should probably be using [`Generics::split_for_impl`](https://docs.rs/syn/latest/syn/struct.Generics.html#method.split_for_impl)

baloo commented:
@tarcieri injecting a default lifetime with `split for impl` in wasn't trivial but this is the best I came up with. I'd love your opinion

baloo commented:
rebased on #1014

baloo commented:
So this needs:
 - [ ] release of `der_derive` to `0.7.1`
 - [ ] `der` needs to bump dependency to `der_derive` to `0.7.1`
 - [ ] release of `der` (`0.7.4`)
 - [ ] bump `der` here to `0.7.4`.

baloo commented:
alright so that solved that.

lumag commented:
@baloo this now breaks building certificates:
```
error: cannot construct `TbsCertificateInner` with struct literal syntax due to private fields
   --> src/cert_gen.rs:105:43
    |
105 |     let tbs_certificate: TbsCertificate = TbsCertificate {
    |                                           ^^^^^^^^^^^^^^
    |
    = note: ... and other private field `_profile` that was not provided

error: could not compile `qcom-sig-tools` due to previous error

```

tarcieri commented:
Oof, that seems bad. That's another argument for keeping the generic parameters off every type except `Certificate`

lumag commented:
Exporting the `PhantomData` should work

tarcieri commented:
yes, albeit a bit ugly to have `PhantomData` in a public API

lumag commented:
For me this is not a big problem, since I will most likely rewrite my code to use `CertificateBuilder`. But you should at least be aware of thi problem.

baloo commented:
We can either revert the thing for now and rework that, or get a constructor (`new`) for the `TbsCertificate` and the `Certificate.

baloo commented:
or we just get rid of the phantom data there ... https://github.com/RustCrypto/formats/pull/1017
